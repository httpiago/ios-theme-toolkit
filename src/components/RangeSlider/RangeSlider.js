import React from 'react'
import PropTypes from 'prop-types'
import { mapClassList } from '../../Utils'

/**
 * RangeSlider component.
 *
 * @since 0.1.0
 * @see https://httpiago.github.io/ios-theme-toolkit/#/components/RangeSlider
 *
 * @todo Quandos o React hooks for lançado fazer um componente controlado para criar uma barrinha colorida atrás da bolinha do elemento e um balãozinho com o valor selecionado em cima .
 */
export default function RangeSlider({ block, /*color,*/ className: aditionalClasses, ...rest }) {

  const prefix = 'range-slider'
  const classes = mapClassList({
    // [`${prefix}--${color}`]: color,
    [`${prefix}--block`]: block,
  })

  return (
    <input
      type="range"
      className={`${prefix} ${classes} ${aditionalClasses}`}
      {...rest}
    />
  )
}

RangeSlider.defaultProps = {
  className: '',
  // color: 'default',
  disabled: false,
  block: false,
  min: 0,
  max: 100,
  step: 1,
  onChange: function(event) {}
}

RangeSlider.propTypes = {
  /** Aditional classes. */
  className: PropTypes.string,
  /** Minimum value. */
  min: PropTypes.number,
  /** Maximum value. */
  max: PropTypes.number,
  /** Minimal step between values. */
  step: PropTypes.number,
  /** The input value for CONTROLLED components */
  value: PropTypes.number,
  /** The initial input value for UNCONTROLLED components */
  defaultValue: PropTypes.number,
  /** Color of the element. */
  // color: PropTypes.oneOf([ 'default', 'blue', 'red', 'green', 'yellow', 'gray', 'white' ]),
  /** Disable or not the range. */
  disabled: PropTypes.bool,
  /** Fit range width to its parent width. */
  block: PropTypes.bool,
  /** Event fired when slider value change. */
  onChange: PropTypes.func
}
