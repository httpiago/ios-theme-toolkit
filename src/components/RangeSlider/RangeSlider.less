@import '../../customizations.less';

.range-slider {
  -webkit-appearance: none;
  appearance: none;
  cursor: pointer;
  position: relative;
  padding: 0;
  margin: 0;

  .range-slider__thumb() {
    background-color: #fff;
    border: none;
    -webkit-box-shadow: 0 2px 4px rgba(0,0,0,.3);
    box-shadow: 0 2px 4px rgba(0,0,0,.3);
    -webkit-box-sizing: border-box;
    box-sizing: border-box;
    border-radius: 50%;
    position: absolute;
    height: 26px;
    width: 26px;
    margin-top: -13px;
    margin-left: -13px;
    z-index: 1;
    -webkit-appearance: none;
    appearance: none;
  }

  &::-webkit-slider-thumb {
    .range-slider__thumb()
  }
  &::-moz-range-thumb{
    .range-slider__thumb()
  }
  &::-ms-thumb{
    .range-slider__thumb()
  }

  .range-slider__track() {
    background: #b7b8b7;
    position: relative;
    height: 1px;
    width: 100%;
    overflow: hidden;
    left: 0;
    top: 50%;
    -webkit-appearance: none;
    appearance: none;
  }

  &::-webkit-slider-runnable-track {
    .range-slider__track()
  }
  &::-moz-range-track {
    .range-slider__track()
  }
  &::-ms-track {
    .range-slider__track()
  }
}

// Disabled range
.range-slider[disabled] {
  opacity: .6;
  pointer-events: none;
  cursor: default;
}

// Block range
.range-slider--block {
  width: 100%;
  margin: 6px 0;
}
