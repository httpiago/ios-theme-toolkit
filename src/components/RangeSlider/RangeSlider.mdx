---
name: RangeSlider
route: /components/RangeSlider
menu: Componentes
---

import { Playground, PropsTable } from 'docz'
import { RangeSlider } from '../../index.js'
import '../../../dist/styles.css'

# RangeSlider

To use this component, simply import it into any file and pass the [modifiers](#/#props) in the properties.

```jsx
import { RangeSlider } from 'ios-theme-toolkit'

export default () => (
  <RangeSlider value={30} />
)
```

## Examples

<Playground>
  <span className="desc">Default:</span>
  <RangeSlider min={0} max={50} defaultValue={25} />

  <span className="desc">Range that occupies the entire horizontal space:</span>
  <RangeSlider block min={0} max={100} defaultValue={30} />

  <span className="desc">Disabled:</span>
  <RangeSlider disabled defaultValue={75} />

  <style>{".react-live-preview .desc { display: block; font-weight: bold; margin-top: 20px }"}</style>
</Playground>

## Props

<PropsTable of={ RangeSlider } />

All the properties which are not listed above will be transferred to the <\input type="range"\> tag, so, It accepts all props which native input support.
