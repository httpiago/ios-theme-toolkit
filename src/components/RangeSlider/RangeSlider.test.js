/**
 * TEST FILE
 * Please try writing the tests for all component behaviors,
 * manually call the functions within the class, change the properties...
 *
 * @see https://jestjs.io/docs/en/expect
 * @see https://github.com/airbnb/enzyme/tree/master/docs/api
 */
import React from 'react'
import { shallow } from 'enzyme'
import RangeSlider from './RangeSlider'

const COMPONENT = shallow(
  <RangeSlider value={42} onChange={() => {}} className="aditionalClass" />
)

describe('Test RangeSlider component', () => {
  it('should render correctly based on last snapshot', () => {
    expect(COMPONENT).toMatchSnapshot()
  })

  it('should update correctly', () => {
    COMPONENT.setProps({ value: 76 })
    
    expect(COMPONENT.html()).toContain('value="76"')
  })

  it('should has aditional classes', () => {
    expect(COMPONENT.hasClass('aditionalClass')).toBeTruthy()
  })

  it('should pass aditional props to html element', () => {
    COMPONENT.setProps({ 'data-custom-attr': 'yes' })

    expect(COMPONENT.html()).toContain('data-custom-attr="yes"')
  })

  it('should return a input[type="range"] tag', () => {
    expect(COMPONENT.type()).toEqual('input')

    expect(COMPONENT.html()).toContain('type="range"')
  })

})
