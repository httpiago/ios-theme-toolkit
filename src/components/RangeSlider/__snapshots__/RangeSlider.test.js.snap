// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test RangeSlider component should render correctly based on last snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RangeSlider
    block={false}
    className="aditionalClass"
    disabled={false}
    max={100}
    min={0}
    onChange={[Function]}
    step={1}
    value={42}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "className": "range-slider  aditionalClass",
      "disabled": false,
      "max": 100,
      "min": 0,
      "onChange": [Function],
      "step": 1,
      "type": "range",
      "value": 42,
    },
    "ref": null,
    "rendered": null,
    "type": "input",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "className": "range-slider  aditionalClass",
        "disabled": false,
        "max": 100,
        "min": 0,
        "onChange": [Function],
        "step": 1,
        "type": "range",
        "value": 42,
      },
      "ref": null,
      "rendered": null,
      "type": "input",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
